tests/sources/test_pyproject_setuptools.py

```diff
@@ -0,0 +1,355 @@
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """))
+    result = update_python_requires(str(filename), v(['3.5', '3.6', '3.7']))
+    assert result is not None
+    assert "\n".join(result) == textwrap.dedent("""\
+        [project]
+            name='foo'
+            requires-python = ">=3.5"
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """)
+
+
+def test_update_python_requires_file_object(fix_max_python_3_version):
+    fix_max_python_3_version(7)
+    fp = StringIO(textwrap.dedent("""\
+        [project]
+            name='foo'
+            requires-python = ">=3.4"
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['3.5', '3.6', '3.7']))
+    assert result is not None
+    assert "\n".join(result) == textwrap.dedent("""\
+        [project]
+            name='foo'
+            requires-python = ">=3.5"
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """)
+
+
+def test_update_python_requires_when_missing(capsys):
+    fp = StringIO(textwrap.dedent("""\
+        [project]
+            name='foo'
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['3.5', '3.6', '3.7']))
+    assert result is None
+    assert capsys.readouterr().err == ""
+
+
+def test_update_python_requires_preserves_style(fix_max_python_3_version):
+    fix_max_python_3_version(2)
+    fp = StringIO(textwrap.dedent("""\
+        [project]
+            name='foo'
+            requires-python = ">=2.7,!=3.0.*"
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['2.7', '3.2']))
+    assert "\n".join(result) == textwrap.dedent("""\
+        [project]
+            name='foo'
+            requires-python = ">=2.7,!=3.0.*,!=3.1.*"
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """)
+
+
+def test_update_python_requires_multiline_error(capsys):
+    fp = StringIO(textwrap.dedent("""\
+        [project]
+            name='foo'
+            requires-python = '>=2.7, !=3.0.*'
+        [build-system]
+            requires = ["setuptools", "setuptools-scm"]
+            build-backend = "setuptools.build_meta"
+    """))
+    fp.name = "pyproject.toml"
+    result = update_python_requires(fp, v(['2.7', '3.2']))
+    assert result == ['[project]',
```

Comment:
This is wrong: the FileLines returned by update_... functions are supposed to include trailing newlines.